### PR TITLE
fix: password timing attack

### PR DESCRIPTION
**Pull request by mmooyyii**
Possible risks in the code for checking passwords. 

use a constant-time algorithm.

There is more information in  [https://codahale.com/a-lesson-in-timing-attacks](https://codahale.com/a-lesson-in-timing-attacks)

### CODE REVIEW

File: apps/emqx/src/emqx_passwd.erl [hunk 1]
```diff
@@ -51,7 +51,12 @@ check_pass({PassHash, Salt, Password}, {salt, HashType}) ->
     check_pass(PassHash, emqx_passwd:hash(HashType, <<Salt/binary, Password/binary>>));
 check_pass({PassHash, Salt, Password}, {HashType, salt}) ->
     check_pass(PassHash, emqx_passwd:hash(HashType, <<Password/binary, Salt/binary>>));
-check_pass(PassHash, PassHash) -> ok;
+check_password(PassHash1, PassHash2) when size(PassHash1) =:= size(PassHash2) ->
```
Comment (k32) [inline apps/emqx/src/emqx_passwd.erl:54]: This code won't compile because the clause name is different.

File: apps/emqx/src/emqx_passwd.erl [hunk 2]
```diff
@@ -51,7 +51,12 @@ check_pass({PassHash, Salt, Password}, {salt, HashType}) ->
...
+    Zip = lists:zip(binary_to_list(PassHash1), binary_to_list(PassHash2)),
+    case lists:foldl(fun({C1, C2}, Acc) -> Acc bor (C1 bxor C2) end, 0, Zip) of
+        0 -> ok;
+        _ -> {error, password_error}
+    end.
```
Comment (k32) [inline apps/emqx/src/emqx_passwd.erl:59]: Please factor out this code to a separate function and create a unit tests for it.

File: apps/emqx/src/emqx_passwd.erl (hunk 1 above)
Comment (k32) [inline apps/emqx/src/emqx_passwd.erl:54]: My guess is that the following implementation will be more efficient:
```erlang
Size = bit_size(PassHash1),
case bit_size(PassHash2) of
  Size -> 
    <<N1:Size>> = PassHash1,
    <<N2:Size>> = PassHash2,
    case N1 bxor N2 of
      0 -> ok;
      _ -> {error, password_error}
    end;
  _ -> 
    {error, password_error}
end
```

**Review by k32 — CHANGES_REQUESTED** [review]

File: apps/emqx/src/emqx_passwd.erl (hunk 1 above)
Comment (mmooyyii) [reply apps/emqx/src/emqx_passwd.erl:54]: sorry , I did not open my ide, edit it on web directly.

File: apps/emqx/src/emqx_passwd.erl (hunk 2 above)
Comment (mmooyyii) [reply apps/emqx/src/emqx_passwd.erl:59]: I am not very familiar with this project. just look this pull request as a issue.

**Pull request closed without merging**